### PR TITLE
:hammer_and_wrench: Add Harden Runner to improve CI-security (2/2) (#956)

### PR DESCRIPTION
**Pull request by LogFlames**
Some action are only run when merged into the main branch. This appiles the recommended configuration of harden-runner from those actions.

Some actions will need to be more strictly configured later.

### CODE REVIEW

File: .github/workflows/Lockfile.yml [hunk 1]
```diff
@@ -12,9 +12,18 @@ jobs:
       contents: write
     runs-on: ubuntu-latest
     steps:
-      - uses: step-security/harden-runner@91182cccc01eb5e619899d80e4e971d6181294a7 # v2.10.1
+      - name: Harden Runner
+        uses: step-security/harden-runner@91182cccc01eb5e619899d80e4e971d6181294a7 # v2.10.1
         with:
-          egress-policy: audit
+          disable-sudo: true
+          egress-policy: block
```
Comment (algomaster99) [inline .github/workflows/Lockfile.yml:19]: What the possible options? I could only see `block` and `audit`.

File: .github/workflows/Lockfile.yml [hunk 2]
```diff
@@ -12,9 +12,18 @@ jobs:
...
+          allowed-endpoints: >
+            github.com:443
+            objects.githubusercontent.com:443
+            raw.githubusercontent.com:443
+            repo.maven.apache.org:443
+            repo1.maven.org:443
+            sh.jbang.dev:443
```
Comment (algomaster99) [inline .github/workflows/Lockfile.yml:26]: How are you getting this list of endpoints?

Comment (LogFlames) [reply .github/workflows/Lockfile.yml:26]: In the summary (accessible at the link printed in the harden-runner step) they list all endpoints that were accessed. I used the "Recommended" tab where they contained the recommended config.

Comment (LogFlames) [reply .github/workflows/Lockfile.yml:26]: https://app.stepsecurity.io/github/chains-project/maven-lockfile/actions/runs/11597128578

File: .github/workflows/Lockfile.yml (hunk 1 above)
Comment (LogFlames) [reply .github/workflows/Lockfile.yml:19]: Those are the only possible values:
https://github.com/step-security/harden-runner/blob/main/action.yml

If it's set to block you can choose to not report the results to their API.

**Comment by algomaster99** [comment]
Cool! So first you can audit to record the allowlist and then block it, right?


Regards,
Aman Sharma

PhD Student
KTH Royal Institute of Technology
School of Electrical Engineering and Computer Science (EECS)
Department of Theoretical Computer Science (TCS)
<http://www.kth.se><https://www.kth.se/profile/amansha><https://www.kth.se/profile/amansha>
<https://www.kth.se/profile/amansha>https://algomaster99.github.io/
________________________________

**Comment by LogFlames** [comment]
Yes precisely, that's why I split the PR into two as well. The workflow had to run to collect all endpoints. And the publish workflow is still only in audit state. My plan is to record endpoints and hard-code them when I do the next release, although I might keep it in audit mode since it can be quite difficult to fix if the workflow breaks/is blocked half way through. And it can still be useful to check which endpoints were connected to afterwards - It would not prevent a leak but then you can see that you need to rotate the tokens etc

I am missing some option to make an allowlist for file writes as well. Harden-runner currently reports that there are for example 5 file write operations and list the files. But I want to have it report especially if those files changes. Now it's infeasible to check each workflow-run and compare which files were written to.

**Comment by LogFlames** [comment]
But it was very easy to add to the project!

And they have a community version which is free for public repos. It doesn't include all features but do the basic and blocks un-allowlisted connections.

They can also create a "security pr" with best practises where they e.g. add depandabot, and setup some other recommended pre-commit hooks etc. I looked through it and dont think it applies to this project, but it could be an easy way to add good defaults.

**Comment by LogFlames** [comment]
Although it seems some endpoints are not connected to all the time. For example the repo.maven.apache.org:443 was new for a secondary run (https://app.stepsecurity.io/github/chains-project/maven-lockfile/actions/runs/11600602474?jobid=32301416922&tab=recommendations). It would probably be a good idea to let it run a couple of times and connect endpoints, but I thought github actions would be deterministic :thinking:

**Comment by algomaster99** [comment]
> For example the repo.maven.apache.org:443 was new for a secondary run ([app.stepsecurity.io/github/chains-project/maven-lockfile/actions/runs/11600602474?jobid=32301416922&tab=recommendations](https://app.stepsecurity.io/github/chains-project/maven-lockfile/actions/runs/11600602474?jobid=32301416922&tab=recommendations)). It would probably be a good idea to let it run a couple of times and connect endpoints, but I thought github actions would be deterministic

That is very strange :open_mouth: 
Can we know the exact endpoint with path? Why was it queried only in the second run?

**Pull request merged**